File: 字符串/删除字符使频率相同.java
package 字符串;

import java.util.HashMap;
import java.util.Map;

//必须要删除一个字符，使得所有字符出现的次数相同，给定的字符串能满足这个条件吗
//88场双周赛  差点没写出来
public class 删除字符使频率相同 {
    int[] fre = new int[26];

    public boolean equalFrequency(String word) {
        for (int i = 0; i < word.length(); i++) {
            fre[word.charAt(i) - 'a']++;
        }
        //映射频率   -》 该频率的字母有多少个
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < 26; i++) {
            if (fre[i] == 0) continue;
            map.put(fre[i], map.getOrDefault(fre[i], 0) + 1);
        }
        if (map.size() == 1) {
            for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
                //这里差点没写的出来
                return entry.getKey() == 1 || entry.getValue() == 1;
            }
        }
        if (map.size() == 2) {
            int pre = -1;
            int preFre = 0;
            int now = 0;
            int nowFre = 0;
            for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
                if (pre == -1) {
                    pre = entry.getValue();
                    preFre = entry.getKey();
                } else {
                    now = entry.getValue();
                    nowFre = entry.getKey();
                }
            }
            if (preFre == 1) {
                if (pre == 1) return true;
            }
            if (nowFre == 1) {
                if (now == 1) return true;
            }
            if (nowFre - preFre == 1 && now == 1) return true;
            if (preFre - nowFre == 1 && pre == 1) return true;
        }
        return false;
    }
}

File: 字符串/删除字符使频率相同.c
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#include "/Users/zhui/IdeaProjects/leetcode/uthash/include/uthash.h"

/***
 * hashtable的实现
 * */

typedef struct {
    int key;
    int val;
    UT_hash_handle hh;
} HashItem;

HashItem* hashFindItem(HashItem** obj, int key) {
    HashItem* pEntry = NULL;
    HASH_FIND_INT(*obj, &key, pEntry);
    return pEntry;
}

bool hashAddItem(HashItem** obj, int key, int val) {
    if (hashFindItem(obj, key)) {
        return false;
    }
    HashItem* pEntry = (HashItem*)malloc(sizeof(HashItem));
    pEntry->key = key;
    pEntry->val = val;
    HASH_ADD_INT(*obj, key, pEntry);
    return true;
}

bool hashSetItem(HashItem** obj, int key, int val) {
    HashItem* pEntry = hashFindItem(obj, key);
    if (!pEntry) {
        hashAddItem(obj, key, val);
    } else {
        pEntry->val = val;
    }
    return true;
}

int hashGetItem(HashItem** obj, int key, int defaultVal) {
    HashItem* pEntry = hashFindItem(obj, key);
    if (!pEntry) {
        return defaultVal;
    }
    return pEntry->val;
}

void hashEraseItem(HashItem** obj, const char* key) {
    HashItem* pEntry = NULL;
    HASH_FIND_INT(*obj, &key, pEntry);
    HASH_DEL(*obj, pEntry);
    free(pEntry);
}

void hashFree(HashItem** obj) {
    HashItem *curr = NULL, *tmp = NULL;
    HASH_ITER(hh, *obj, curr, tmp) {
        HASH_DEL(*obj, curr);
        free(curr);
    }
}

bool equalFrequency(char* word) {
    int cnt[26] = {0};
    int n = strlen(word);
    for (int i = 0; i < n; i++)
        cnt[word[i] - 'a']++;

    // 映射频率   -》 该频率的字母有多少个
    HashItem* freqCount = NULL;
    for (int i = 0; i < 26; i++) {
        if (cnt[i] == 0)
            continue;

        hashSetItem(&freqCount, cnt[i], hashGetItem(&freqCount, cnt[i], 0) + 1);
    }

    HashItem* tmp = NULL;
    if (HASH_COUNT(freqCount) == 1) {
        int key , value;
        for (tmp = freqCount; tmp != NULL; tmp = (HashItem*)tmp->hh.next) {
            key = tmp->key;
            value = tmp->val;
        }
        hashFree(&freqCount);
        return key == 1 || value == 1; 
    }

    if (HASH_COUNT(freqCount) == 2) {
        int pre = -1;
        int preFre = 0;
        int now = 0;
        int nowFre = 0;
        for (tmp = freqCount; tmp != NULL; tmp = (HashItem*)tmp->hh.next) {
            if (pre == -1) {
                pre = tmp->val;
                preFre = tmp->key;
            } else {
                now = tmp->val;
                nowFre = tmp->key;
            }
        }
        if (preFre == 1 && pre == 1) {
            hashFree(&freqCount);
            return true;
        }
        if (nowFre == 1 && now == 1) {
            hashFree(&freqCount);
            return true;
        }
        if (nowFre - preFre == 1 && now == 1) {
            hashFree(&freqCount);
            return true;
        }
        if (preFre - nowFre == 1 && pre == 1) {
            hashFree(&freqCount);
            return true;
        }
    }

    hashFree(&freqCount);
    return false;
}

int main() {
    bool res = equalFrequency("abcc");
    printf("%u\n", res);
}